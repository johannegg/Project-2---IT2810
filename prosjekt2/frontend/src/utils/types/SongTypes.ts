export type Artist = {
    id: string;
    name: string;
};

export type Genre = {
    name: string;
};

export type SongData = {
    id: string;
    title: string;
    views: number;
    year: number;
    artist: Artist;
    genre: Genre;
    lyrics: string;
};