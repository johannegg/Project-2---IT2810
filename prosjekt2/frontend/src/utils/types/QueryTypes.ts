import { SongData } from "./SongTypes";

// Define types for the GraphQL response
export type SongsQueryResponse = {
    songs: SongData[];
};