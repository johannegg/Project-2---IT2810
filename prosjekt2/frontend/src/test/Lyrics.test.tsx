import React from "react";
import { describe, test, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SongData } from "../utils/types/SongTypes";
import Lyric from "../components/Lyrics/Lyrics";

vi.mock("../components/FavoriteButton/FavoriteButton", () => ({
    default: () => <button>Favorite</button>,
}));

vi.mock("../components/PlusMinusButton/PlusMinusButton", () => ({
    default: () => <button>Plus/Minus</button>,
}));

vi.mock("../components/BackButton/BackButton", () => ({
    default: () => <button>Back</button>,
}));

vi.mock("../utils/FormatViews", () => ({
    formatViews: (views: number) => `${views.toLocaleString()}`,
}));

describe("Lyric Component", () => {
    const mockSongData: SongData = {
        id: "1",
        title: "Test Song",
        views: 12345,
        year: 2022,
        artist: { id: "1", name: "Test Artist" },
        genre: { name: "Pop" },
        lyrics: "This is line 1\nThis is line 2\nThis is line 3",
    };

    // Snapshot Tests
    test("matches snapshot", () => {
        const { container } = render(<Lyric songData={mockSongData} />);
        expect(container).toMatchSnapshot();
    });

    // Functional Tests
    test("renders the Lyric component correctly", () => {
        render(<Lyric songData={mockSongData} />);

        // Verify headers and song information
        expect(screen.getByRole("heading", { name: /test song/i })).toBeInTheDocument();
        expect(screen.getByRole("heading", { name: /test artist/i })).toBeInTheDocument();
        expect(screen.getByText(/release year: 2022/i)).toBeInTheDocument();
        expect(screen.getByText(/genre: pop/i)).toBeInTheDocument();
        expect(screen.getByText(/12,345 views/i)).toBeInTheDocument();

        // Verify lyrics rendering
        expect(screen.getByText(/this is line 1/i)).toBeInTheDocument();
        expect(screen.getByText(/this is line 2/i)).toBeInTheDocument();
        expect(screen.getByText(/this is line 3/i)).toBeInTheDocument();

        // Verify the presence of buttons
        expect(screen.getByRole("button", { name: /back/i })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: /favorite/i })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: /plus\/minus/i })).toBeInTheDocument();
    });

    test("formats views correctly", () => {
        render(<Lyric songData={mockSongData} />);

        // Verify formatted views
        expect(screen.getByText(/12,345 views/i)).toBeInTheDocument();
    });

    test("renders lyrics with line breaks", () => {
        render(<Lyric songData={mockSongData} />);

        // Verify line breaks in lyrics
        const lines = screen.getAllByText(/this is line/i);
        expect(lines).toHaveLength(3); 
    });

    test("interacts with buttons", async () => {
        const user = userEvent.setup();
        render(<Lyric songData={mockSongData} />);

        // Simulate button clicks
        const backButton = screen.getByRole("button", { name: /back/i });
        const favoriteButton = screen.getByRole("button", { name: /favorite/i });
        const plusMinusButton = screen.getByRole("button", { name: /plus\/minus/i });

        await user.click(backButton);
        await user.click(favoriteButton);
        await user.click(plusMinusButton);

        // Verify buttons are clickable (mocked components handle internal logic)
        expect(backButton).toBeEnabled();
        expect(favoriteButton).toBeEnabled();
        expect(plusMinusButton).toBeEnabled();
    });
});