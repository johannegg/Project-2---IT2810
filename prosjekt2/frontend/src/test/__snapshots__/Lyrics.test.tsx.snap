// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Lyric Component > matches snapshot 1`] = `
<div>
  <div
    class="lyrics"
  >
    <div
      class="lyrics-header"
    >
      <button>
        Back
      </button>
      <div
        class="buttons-right-container"
      >
        <button>
          Favorite
        </button>
        <button>
          Plus/Minus
        </button>
      </div>
    </div>
    <h1>
      Test Song
    </h1>
    <h2>
      Test Artist
    </h2>
    <section
      class="songInfo"
    >
      <p>
        Release year: 
        2022
      </p>
      <p>
        Genre: 
        Pop
      </p>
      <div
        class="viewsInfo"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          style="margin-right: 5px;"
          viewBox="0 0 576 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
          />
        </svg>
        <p>
          12,345
           Views
        </p>
      </div>
    </section>
    <hr />
    <p>
      <span>
        This is line 1
        <br />
      </span>
      <span>
        This is line 2
        <br />
      </span>
      <span>
        This is line 3
        <br />
      </span>
    </p>
  </div>
</div>
`;
